File: backend/app/routers/students.py
import os
import uuid
from datetime import date, datetime

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from sqlalchemy import and_, delete, func, or_, select
from sqlalchemy.orm import Session

from ..config import settings
from ..core.deps import require_staff
from ..db import get_db
from ..models.post_lesson import PostLessonPlan
from ..models.post_student import PostStudent
from ..schemas.common import Page
from ..schemas.student import StudentCreate, StudentOut, StudentUpdate
from ..services import registration_number as regsvc
from ..services.student_service import calc_age, create_default_lesson_plans

router = APIRouter(
    prefix="/api/students",
    tags=["students"],
    dependencies=[Depends(require_staff)],
)

FUNNEL_CONDS = {
    "confession": lambda: PostStudent.confession_date.isnot(None),
    "baptism": lambda: PostStudent.baptism_date.isnot(None),
    "gotochurch": lambda: PostStudent.send_to_church_date.isnot(None),
    "serving": lambda: and_(PostStudent.serving.isnot(None), PostStudent.serving != ""),
}


@router.get("/next-code")
def next_code(type: str = "individual", db: Session = Depends(get_db)):
    if type == "network":
        return {"registration_number": regsvc.next_network(db, date.today())}
    return {"registration_number": regsvc.next_individual(db)}


@router.get("", response_model=Page[StudentOut])
def list_students(
    db: Session = Depends(get_db),
    page: int = 1,
    per_page: int = 20,
    search: str | None = None,
    province: str | None = None,
    funnel: str | None = None,
):
    conds = []
    if search:
        like = f"%{search}%"
        conds.append(
            or_(
                PostStudent.registration_number.like(like),
                PostStudent.first_name.like(like),
                PostStudent.last_name.like(like),
                PostStudent.phone_number.like(like),
            )
        )
    if province:
        conds.append(PostStudent.province == province)
    if funnel in FUNNEL_CONDS:
        conds.append(FUNNEL_CONDS[funnel]())

    base = select(PostStudent)
    count_q = select(func.count()).select_from(PostStudent)
    for c in conds:
        base = base.where(c)
        count_q = count_q.where(c)

    total = db.execute(count_q).scalar() or 0
    per_page = min(max(per_page, 1), 100)
    page = max(page, 1)
    rows = db.execute(
        base.order_by(PostStudent.id.desc()).offset((page - 1) * per_page).limit(per_page)
    ).scalars().all()
    pages = (total + per_page - 1) // per_page if per_page else 1
    return Page(items=rows, total=total, page=page, per_page=per_page, pages=pages)


@router.get("/{sid}", response_model=StudentOut)
def get_student(sid: int, db: Session = Depends(get_db)):
    s = db.get(PostStudent, sid)
    if not s:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "ไม่พบนักเรียน")
    return s


@router.post("", response_model=StudentOut, status_code=status.HTTP_201_CREATED)
def create_student(body: StudentCreate, db: Session = Depends(get_db)):
    now = datetime.now()
    data = body.model_dump()
    reg = data.pop("registration_number", None) or regsvc.next_individual(db)
    if not data.get("age") and data.get("birthday_date"):
        data["age"] = calc_age(data["birthday_date"])
    student = PostStudent(registration_number=reg, created_at=now, updated_at=now, **data)
    db.add(student)
    db.flush()
    create_default_lesson_plans(db, student.id)
    db.commit()
    db.refresh(student)
    return student


@router.put("/{sid}", response_model=StudentOut)
def update_student(sid: int, body: StudentUpdate, db: Session = Depends(get_db)):
    student = db.get(PostStudent, sid)
    if not student:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "ไม่พบนักเรียน")
    data = body.model_dump(exclude_unset=True)
    for k, v in data.items():
        setattr(student, k, v)
    if data.get("birthday_date"):
        student.age = calc_age(data["birthday_date"])
    student.updated_at = datetime.now()
    db.commit()
    db.refresh(student)
    return student


@router.delete("/{sid}", status_code=status.HTTP_204_NO_CONTENT)
def delete_student(sid: int, db: Session = Depends(get_db)):
    student = db.get(PostStudent, sid)
    if student:
        db.execute(delete(PostLessonPlan).where(PostLessonPlan.post_student_id == sid))
        db.delete(student)
        db.commit()
    return None


@router.post("/{sid}/testimony", response_model=StudentOut)
async def upload_testimony(
    sid: int, file: UploadFile = File(...), db: Session = Depends(get_db)
):
    student = db.get(PostStudent, sid)
    if not student:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "ไม่พบนักเรียน")
    ext = os.path.splitext(file.filename or "")[1].lower()
    fname = f"{uuid.uuid4().hex}{ext}"
    dest_dir = os.path.join(settings.MEDIA_DIR, "testimonies")
    os.makedirs(dest_dir, exist_ok=True)
    with open(os.path.join(dest_dir, fname), "wb") as fh:
        fh.write(await file.read())
    student.testimony_file = f"testimonies/{fname}"
    student.has_testimony = True
    student.updated_at = datetime.now()
    db.commit()
    db.refresh(student)
    return student

File: frontend/lib/students.ts
import { apiFetch, API_URL } from "./api";

export type Student = {
  id: number;
  registration_number: string;
  registration_date?: string | null;
  first_name: string;
  last_name: string;
  gender?: string | null;
  religion?: string | null;
  address?: string | null;
  tambon?: string | null;
  amphure?: string | null;
  province?: string | null;
  zipcode?: string | null;
  phone_number?: string | null;
  age?: string | null;
  profession?: string | null;
  birthday_date?: string | null;
  church_name?: string | null;
  send_to_church_date?: string | null;
  program?: string | null;
  confession_date?: string | null;
  baptism_date?: string | null;
  serving?: string | null;
  note?: string | null;
  has_testimony?: boolean | null;
  testimony_file?: string | null;
};

export type PageResult<T> = {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
};

export function listStudents(params: {
  page?: number;
  per_page?: number;
  search?: string;
  province?: string;
  funnel?: string;
}) {
  const qs = new URLSearchParams();
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== "") qs.set(k, String(v));
  });
  return apiFetch<PageResult<Student>>(`/api/students?${qs.toString()}`);
}

export const getStudent = (id: number) => apiFetch<Student>(`/api/students/${id}`);

export const createStudent = (data: Partial<Student>) =>
  apiFetch<Student>("/api/students", { method: "POST", body: JSON.stringify(data) });

export const updateStudent = (id: number, data: Partial<Student>) =>
  apiFetch<Student>(`/api/students/${id}`, { method: "PUT", body: JSON.stringify(data) });

export const deleteStudent = (id: number) =>
  apiFetch<void>(`/api/students/${id}`, { method: "DELETE" });

export const nextCode = (type: "individual" | "network" = "individual") =>
  apiFetch<{ registration_number: string }>(`/api/students/next-code?type=${type}`);

// geo cascade
export const getProvinces = () => apiFetch<string[]>("/api/geo/provinces");
export const getAmphures = (province: string) =>
  apiFetch<string[]>(`/api/geo/amphures?province=${encodeURIComponent(province)}`);
export const getTambons = (amphure: string) =>
  apiFetch<{ name_th: string; zip_code: number | null }[]>(
    `/api/geo/tambons?amphure=${encodeURIComponent(amphure)}`,
  );

// lookups
export const getProfessions = () => apiFetch<string[]>("/api/lookups/professions");
export const getReligions = () =>
  apiFetch<{ short_name: string; name: string | null }[]>("/api/lookups/religions");

export async function uploadTestimony(id: number, file: File) {
  const fd = new FormData();
  fd.append("file", file);
  const res = await fetch(`${API_URL}/api/students/${id}/testimony`, {
    method: "POST",
    credentials: "include",
    body: fd,
  });
  if (!res.ok) throw new Error(`upload ${res.status}`);
  return (await res.json()) as Student;
}
